Dev 1.1

rdaitan opened the pull request:
Filter threads according to category.
Search threads, comments, or users. (medium priority)
Follow notifications, follow/unfollow AJAX (low priority)
Pagination

Account
    - Register Account
    - View/Edit Account
    - View other user
    - Login/Logout

Thread
    - Create Thread
    - Edit/Delete own thread
    - View Thread
    - Follow/unfollow thread

Comment
    - Create comment
    - Edit/delete own comment
    - view comment

rdaitan committed:
Created new database configuration

New tables: category, follow.

rdaitan committed:
Changed DB configuration.

GRANT command has IDENTIFIED by which I jsut found out to be the
password. Everytime I reconfigure the database, I must change the
password and I just found out why.

Also inserted categories.

rdaitan committed:
Fixed spelling of CREATE_THREAD_URL constant

rdaitan committed:
Merge branch 'url_constants_fix' into dev_1.1

rdaitan committed:
Updated old db configuration

rdaitan committed:
Added comment methods

Comment::get()
- Retrieve a comment by id. Return false if no record was found.

Comment::update()
- Update this instance of a comment with its new values.

rdaitan committed:
Comment edit

rdaitan committed:
Comment view

rdaitan committed:
Modified db configuration

Insert initial values.

rdaitan committed:
Merge branch 'issue#1' into dev_1.1

rdaitan committed:
Comment delete

rdaitan committed:
Redo comment edit

Added exception handler to handle exceptions without having to rely
on try catch.

Added Comment::getOrFail() which throws a RecordNotFoundException if
a record was not found.

rdaitan committed:
Code cleanup

rdaitan committed:
Merge branch 'issue#2' into dev_1.1

rdaitan committed:
Added category select in thread create view

Removed username validation message html

It was in the thread create view. I don't need it anymore but I
forgot to remove it.

Added category select field

rdaitan committed:
Modified thread controller

Retrieved categories from category table. Stored category to database.
No validation for category yet. I suppose, db would throw an exception
because of a constraint failure. I'll take advantage of that.

rdaitan committed:
Created category model

Contains getAll() static function to retrieve all the categories.

rdaitan committed:
Updated thread creation to store category id

Saved category id to thread table

Thread::create() throws CategoryException

When the category input does not match any category, a CategoryException
will be thrown. The constant, Thread::ERR_CATEGORY is actually the
code for a foreign key constraint failure. However, in the Thread
model, that kind of failure can only exist from the category_id column,
so I named it as such.

rdaitan committed:
Validate category id

Added validation error on undefined category

Show category validation error

rdaitan committed:
Show edit thread link on own threads
Show edit button on own threads

Used Thread::isOwnedBy() to check for ownership

This is so I won't have to modify the controller and add another variable
to the the thread instance

Add method isOwnedBy()

Checks if the thread is owned by the specified user.

rdaitan committed:
Added ThreadController::edit()
Added Comment::getFirstInThread()

The thread body or the first comment in the thread.

Created thread edit form

A copy-paste of create.php

Created ThreadController::edit()

rdaitan committed:
Added Thread::update() method

rdaitan committed:
Added delete thread link in view

rdaitan committed:
Created delete confirmation page

rdaitan committed:
Added ThreadController::delete()

Shows a confirmation page and deletes the thread when the user confirms.

rdaitan committed:
Added Thread::delete()

rdaitan committed:
Added filter links in thread list

rdaitan committed:
Modified ThreadController::index() to use filters

Results of Thread::getAll() will be filtered and Thread::countAll()
will also be filtered.

rdaitan committed:
Added filtering in Thread methods

Resuls of Thread::getAll() and Thread::countAll() can now have
optional filters.

Added default value to filter parameter

rdaitan committed:
Merge branch 'issue#3' into dev_1.1

rdaitan committed:
Created rank view

Shows the trending threads of the day. Trending threads are the ones
with the most comments made today. Only displays up to 10 threads.

rdaitan committed:
Rank threads by comments created today

rdaitan committed:
Updated view to display comment count

rdaitan committed:
Added Thread::getTrending()

Returns an array of threads based on the ids returned by Comment::
getTrendingThreadIds(). Adds the count of comments created today.

Also displayed trending threads in index sidebar.

rdaitan committed:
Fixes and formatting

Renamed urls.php to url.php

Cleanup

Removed unecessary try...catch

Removed unecessary render()

Removed unecessary code

Refactoring for consistency

Fixed SQL

Cleanup

Cleanup

Formatting

rdaitan committed:
Delete confirmation view

rdaitan committed:
Modified ThreadController::delete()

Now shows the confirmation page.

rdaitan committed:
Add delete links for comment

rdaitan committed:
Merge branch 'issue#4' into dev_1.1

rdaitan committed:
Added follow and unfollow links in thread view

Follow shows up when user is not following a thread, and unfollow
shows up when usser is already following it.

Added text in follow/unfollow link

Defined url constants

For follow.

Fixed follow links

rdaitan committed:
Created FollowController

With add()

Fixed redirect url

rdaitan committed:
Created Follow model

With create() method.

Added methods to check if user follows a thread

rdaitan committed:
Fixed handling of duplicate follows

Also fixed redirection with query.

rdaitan committed:
Fixed undefined object

A guest user sees an undefined object notice when viewing a thread.

rdaitan committed:
Added unfollow action in follow controller

FollowController::remove().

rdaitan committed:
Added Follow::remove()

Also fixed the view again. This is what I get for not testing
before commiting.

rdaitan committed:
Auto-follow thread when created

Thread created by the user is automatically followed.

rdaitan committed:
Fixed ranking

Order of items are not correct.
Now, Comment::countToday() will return an array of the count of
today's comments in each thread. Thread::getTrending() will return
an array of threads based on the array returned by Comment::countToday().
It will also sort the threads based on the number of new comments
and creation date.

rdaitan committed:
Created follow list view

A simple list of all the threads being followed by the user.

rdaitan committed:
Added list all action in follow controller

rdaitan committed:
Added Follow::getAll()

Returns an array of follows by a specified user.

rdaitan committed:
Merge branch 'issue#5' into dev_1.1

rdaitan committed:
Modified user/create view

Added first and last name fields and validation messages.

rdaitan committed:
Modified user/create action

Added first and last name to comment instance.

rdaitan committed:
User::create() stored first and last name

rdaitan committed:
Added validation helper function

validate_name() returns true if the string contains only aplhabetic
characters and/or spaces.

rdaitan committed:
Fixed user validation exceptions

Now differentiates between duplicate username and email.
Message of exception tells which column has a duplicate.

rdaitan committed:
Used getMessage() to access exception error message

rdaitan committed:
Added validation message

For duplicate email

rdaitan committed:
Added user view

rdaitan committed:
Added user/view action

rdaitan committed:
Fixed url

rdaitan committed:
Modified Thread::getAll()

Now accepts different filters.

rdaitan committed:
Modified user/view action

rdaitan committed:
Merge branch 'view-profile' into db-reconfiguration

rdaitan committed:
Added user_id column in thread table

rdaitan committed:
Merge branch 'db-reconfiguration' into view-profile

rdaitan committed:
Added User::getOrFail()

rdaitan committed:
Show authenticated user if user/view does not have query

rdaitan committed:
Revert "Modified Thread::getAll()"

This reverts commit 8a2dee20e79b8088befa475d585f754ad6ec1f4e.

rdaitan committed:
Revert "Modified user/view action"

This reverts commit df908fc03daece23810ab1811b8edcfc0238fa08.

rdaitan committed:
Retrieve threads and comments by user

rdaitan committed:
Fixes

Squashed commit of the following:

commit c45bdafa5775a2f694a732654104cebbf6e0350d
Author: rdaitan <r.daitan@klab.com>
Date:   Tue Sep 15 15:51:59 2015 +0800

    remove links

commit 72b6846da5d37d08a4d5e48884bca6c606ad6fe3
Author: rdaitan <r.daitan@klab.com>
Date:   Tue Sep 15 15:48:43 2015 +0800

    Formatting

commit 2f1cf0a9e6c947cb4a21276f3ffa8c2f7a79b080
Author: rdaitan <r.daitan@klab.com>
Date:   Tue Sep 15 15:45:43 2015 +0800

    Fix sql

commit 7db228a25d787152da8f16fe932273971c4d3c3a
Author: rdaitan <r.daitan@klab.com>
Date:   Tue Sep 15 15:43:44 2015 +0800

    rewrite isOwnedBy

commit 19d90f877be014e0e08b1327e0ef208597394637
Author: rdaitan <r.daitan@klab.com>
Date:   Tue Sep 15 15:40:39 2015 +0800

    Fixed sorting

commit f474bff4ded7aa0846c98dc715fa213885229127
Author: rdaitan <r.daitan@klab.com>
Date:   Tue Sep 15 15:33:14 2015 +0800

    Sort by thread id

commit 8286ef6a82e6f94c29e430a57db51cd73ddb73c3
Author: rdaitan <r.daitan@klab.com>
Date:   Tue Sep 15 15:36:50 2015 +0800

    Add user_id in thread

rdaitan committed:
Added edit profile view

rdaitan committed:
Added user/edit action

rdaitan committed:
Save user changes

rdaitan committed:
Merge branch 'issue#6' into dev_1.1

rdaitan committed:
Added thread search

Squashed commit of the following:

commit 75a88489583b9aa583f3c6788b40b98571c6b526
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 10:26:49 2015 +0800

    Added default value to type

commit c92d7913eb1a7097b2411fcaeb65b08cfc33da90
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 10:17:43 2015 +0800

    Completed search view

commit 40281b69f74bce02e62040df562faa3b4bfdec9f
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 10:00:08 2015 +0800

    Implemented thread search

    Used an instance of stdClass to store results and total count.

commit 84c3392bcfdc0bee649e31d7bb7a0cc59f806895
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 09:29:23 2015 +0800

    Completed pagination in search controller

    Store value in an object instance.

commit b8e507bf2c628059ad054d4f389a76aa1e7ad562
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 09:21:35 2015 +0800

    Thread search in controller

    Incomplete pagination.

commit cc8833a24c459821b1cbdbf1a70d33077b51a4ce
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 08:57:14 2015 +0800

    Created Search view

    Also included controller

rdaitan committed:
Added comment search

rdaitan committed:
Thread searched used Search library

rdaitan committed:
Added user search

rdaitan committed:
Redirect to site index on empty query

rdaitan committed:
Added pagination

rdaitan committed:
Merge branch 'issue#7' into dev_1.1

rdaitan committed:
Thread/index completed
Squashed commit of the following:

commit 80bf260da01bc4b3c65f166987f01121fde4dcc9
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 16:08:01 2015 +0800

    Completed home

commit a1af9a2416a96cda67e6fcae86aa25de2b4b20bc
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 15:53:20 2015 +0800

    Fixed table names

commit cb3534303579f413807ce44920e2b5e5c24a2d19
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 15:51:15 2015 +0800

    Fixed pagination style

commit c67423f4b28e60d147ecd5e1fe88000b5dceb83b
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 15:35:26 2015 +0800

    fixes

commit 890737600e20930bc7d431dc0675fce89097ec2a
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 15:26:55 2015 +0800

    Index

commit d82dc271406e34973037487d6b9c9bab4164c564
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 13:43:05 2015 +0800

    Fixed navbar

commit 85c420ed1bc79d8389ee3644822d66f39438c20f
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 13:40:01 2015 +0800

    Nav bar

commit f9171165925a1830e3fb2c9b7838a33d39bfc672
Author: rdaitan <r.daitan@klab.com>
Date:   Wed Sep 16 11:41:12 2015 +0800

    Imported Skeleton

rdaitan committed:
Cleanup and fixes

rdaitan committed:
Formatting

rdaitan committed:
Merge branch 'issue#8' into dev_1.1

rdaitan committed:
Changed font

rdaitan committed:
Fixes

rdaitan committed:
Merge branch 'issue#8' into dev_1.1

rdaitan committed:
Fixes

rdaitan committed:
Merge branch 'issue#8' into dev_1.1

rdaitan committed:
fixes

and ui changes in log in

rdaitan committed:
finished log in

rdaitan committed:
Create thread done

rdaitan committed:
Thread view

rdaitan committed:
Edit thread

rdaitan committed:
Comment create

rdaitan committed:
Edit comment

rdaitan committed:
Delete confirmation fixes

maragatchalian commented on app/controllers/follow_controller.php:
```diff
@@ -0,0 +1,65 @@
+            // do nothing.
+        }
+
+        // var_dump($follow); die();
```
Please remove unused codes/test codes. :bow:

rdaitan replied:
Removed na po.

maragatchalian commented on app/controllers/thread_controller.php:
```diff
@@ -3,28 +3,42 @@ class ThreadController extends AppController
 
     /*
      * Show all threads.
      */
```
Please be consistent po in using comments. You used a single line comment po in the other parts of your code. :bow:

rdaitan committed:
User profile

maragatchalian commented on app/controllers/user_controller.php:
```diff
@@ -72,6 +91,67 @@ public function logout()
+            $auth_user->new_password = Param::get('new_password');
+
+
+
```
Please remove extra line breaks :bow:

rdaitan replied:
Done.

maragatchalian commented on app/models/thread.php:
```diff
@@ -38,10 +86,40 @@ public static function get($id)
+
+    public static function getTrending($limit)
+    {
+        // $trends = Comment::getTrendingThreadIds($limit);
```
Same as my comment [here](https://github.com/rdaitan/dc-board/pull/2/files#r39704507) po.

rdaitan replied:
Done po. :bow:

rdaitan committed:
Edit profile

rdaitan committed:
Comment search and other fixes

rdaitan committed:
Comment view

rdaitan committed:
User search

rdaitan committed:
Fixed layout

rdaitan committed:
Removed unused views

rdaitan committed:
Merge branch 'issue#8' into dev_1.1

rdaitan committed:
Fixed profile edit

Can now edit without changing password.

rdaitan committed:
Follow updates view

rdaitan committed:
Follow controller

rdaitan committed:
Follow list updating

rdaitan committed:
Added thread updater

rdaitan committed:
Fixes

rdaitan committed:
Updated html

rdaitan committed:
Merge branch 'issue#9' into dev_1.1

rdaitan committed:
Clean up

https://github.com/rdaitan/dc-board/pull/2#discussion_r39704507
https://github.com/rdaitan/dc-board/pull/2#discussion_r39704652
https://github.com/rdaitan/dc-board/pull/2#discussion_r39704750
https://github.com/rdaitan/dc-board/pull/2#discussion_r39704999

rdaitan committed:
Redirect to login on unauthenticated user/view

If user goes to user/view without the id parameter, he will be
redirected to the login page.

rdaitan committed:
Cleaned up exception handler

Defined error messages, and checked of $exception can be dumped.

rdaitan committed:
Redirect to thread view on comment delete

rdaitan committed:
Fixed password update validation

Now properly states which of the current password or new password is wrong.

fatimaPHP commented on app/config/sql/board1_1.sql:
```diff
@@ -0,0 +1,105 @@
+--
+-- Create database
+--
+DROP DATABASE board1_1;
```
This will cause an error if `board1_1` does not exists :sweat:

rdaitan replied:
Changed it po.

fatimaPHP commented on app/config/sql/board1_1.sql:
```diff
@@ -0,0 +1,105 @@
+    first_name  VARCHAR(30) NOT NULL,
+    last_name   VARCHAR(30) NOT NULL,
+    email       VARCHAR(30) NOT NULL UNIQUE,
+    password    VARCHAR(60),
```
I think there should be `NOT NULL` here po unless password is not required. :bow:

rdaitan replied:
Done po.

darwinnavarro commented on app/app_exception.php:
```diff
@@ -18,3 +18,11 @@ class PageNotFoundException extends AppException
+{
+}
+
+class CategoryException extends AppException
```
You may rename the following exceptions to a more specific one. Also consider adding PHPDoc :smile:
ex.

``` php
/**
 * Exception thrown when invalid category is submitted
 */
class InvalidCategoryException extends AppException
{
}
```

rdaitan replied:
Done po. Thank you. :bow:

fatimaPHP commented on app/config/sql/board1_1.sql:
```diff
@@ -0,0 +1,105 @@
+    user_id     INT UNSIGNED NOT NULL,
+    body        TEXT NOT NULL,
+    created_at  TIMESTAMP NOT NULL DEFAULT 0,
+    edited_at   TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
```
suggestion po for the column name, you can change this to `modified_at` :smile:

rdaitan replied:
I see po. It modified_at sounds better :smile: . Thank you.

fatimaPHP commented on app/config/sql/board1_1.sql:
```diff
@@ -0,0 +1,105 @@
+--
+
+START TRANSACTION;
+INSERT INTO category (name) VALUES ('Anime & Manga');
```
There's no `BEGIN;` after `START TRANSACTION;` po :bow:

fatimaPHP commented on app/controllers/follow_controller.php:
```diff
@@ -0,0 +1,101 @@
+            $follow->create();
+        } catch (PDOException $e) {
+            // do nothing.
+        }
```
I think this try-catch block should be in `$follow->create()` method po

rdaitan replied:
Moved na po. Thank you.

fatimaPHP commented on app/controllers/search_controller.php:
```diff
@@ -0,0 +1,59 @@
+        $search = new stdClass();
+        switch($type) {
+        case self::TYPE_THREAD:
+            $search = Thread::search(
```
Please rename `$search` to `$results`

rdaitan replied:
Done po.

fatimaPHP commented on app/helpers/exception_helper.php:
```diff
@@ -0,0 +1,15 @@
+
+set_exception_handler(
+    function (Exception $exception) {
+        var_dump($exception);
```
remove debug lines

rdaitan replied:
Done po.

fatimaPHP commented on app/models/category.php:
```diff
@@ -0,0 +1,25 @@
+    public static function getAll()
+    {
+        $db = DB::conn();
+        $rows = $db->rows(sprintf('SELECT * FROM %s', self::TABLE_NAME));
```
You can simply type the string po rather than using a function to generate the string. If you're considering micro-optimizations :smile:

rdaitan replied:
I put it as a constant po so I could easily rename the table if needed. :smile:

darwinnavarro replied:
I agree with fatima, if we are to consider micro - optimization , functions (in this case sprintf()) uses more memory compared to string. Also declaring a class constant (self::TABLE_NAME) consume memory. I think its better to use string :smile: 

In future projects you will need to consider performance, starting to practice optimization will be helpful :+1:

rdaitan replied:
I see po. I'll change it. :bow:

rdaitan committed:
Moved password update logic

Moved to User::update();

rdaitan committed:
More specific exceptions

CategoryException renamed to InvalidCategoryException.

rdaitan committed:
Updated sql

Added not null in password column.
Renamed edited_at column.
Updated usage.

fatimaPHP commented on app/models/comment.php:
```diff
@@ -47,6 +84,73 @@ public static function getAll($thread_id, $offset, $limit)
+    public static function getFirstInThread(Thread $thread)
+    {
+        $db     = DB::conn();
+        $row    = $db->row(sprintf('SELECT * FROM %s WHERE thread_id=? LIMIT 0, 1', self::TABLE_NAME), array($thread->id));
```
There no need to use LIMIT here po as `row()` is only getting the first record it found from the result set :bow:
Apply to all po

rdaitan replied:
done po.

fatimaPHP commented on app/models/comment.php:
```diff
@@ -54,11 +158,51 @@ public function create(Thread $thread)
+        $db->query(sprintf('DELETE FROM %s WHERE id=?', self::TABLE_NAME), array($this->id));
+    }
+
+    public function isOwnedBy($user)
```
Since the return is boolean po, please rename to ie. isOwner() or isAuthor() :bow:

rdaitan replied:
Done po. :bow:

rdaitan committed:
Moved try catch block to Follow::create()

rdaitan committed:
Renamed $search variable

rdaitan committed:
Renamed isOwnedBy()

rdaitan committed:
Merge branch 'dev_1.1-code_review' into dev_1.1

fatimaPHP commented on app/models/comment.php:
```diff
@@ -47,6 +84,73 @@ public static function getAll($thread_id, $offset, $limit)
+        $row    = $db->row(sprintf('SELECT * FROM %s WHERE thread_id=? ORDER BY id DESC LIMIT 0, 1', self::TABLE_NAME), array($thread->id));
+
+        return $row ? new self($row) : false;
+    }
```
Suggestion po for getting the last comment in a thread, I think it is better to save the ID of the latest comment in thread table po, you would need to have a new column there ie. last_comment_id

With that po you do not need to get all the comments from a thread, sort it, only to get 1 row. It will only take time and slow the system. An extra work is needed though, since you would need to save the last comment ID in thread table every time a new comment is created. :bow:

fatimaPHP replied:
This is just a suggestion po. If you do not have the time to do this, then it's fine po :smile:

rdaitan replied:
I only use this function to get the id of the last comment in thread. What if I just return the id of the last comment with `SELECT MAX(id) FROM comment WHERE thread_id=?`? Will that be fine as well? Thank you for all the suggestions. :bow:

rdaitan committed:
Added title in follows view

rdaitan committed:
Return only id of last comment in thread

rdaitan committed:
Clean up

rdaitan committed:
Table names in string

rdaitan committed:
Clean up

rdaitan committed:
Merge branch 'dev_1.1-code_review' into dev_1.1

rdaitan committed:
Formatting

rdaitan committed:
Reduced whitespace

rdaitan committed:
No more empty comment bodies

rdaitan committed:
Title in comment view

rdaitan committed:
Changed switch to if

rdaitan committed:
Prevented space-only thread body

rdaitan committed:
Merge branch 'dev_1.1-code_review' into dev_1.1

rdaitan committed:
Decreased lines

rdaitan committed:
Adhered to coding standards

rdaitan committed:
Merge branch 'dev_1.1-code_review' into dev_1.1

rdaitan committed:
Hide follow when not logged in

rdaitan committed:
Merge branch 'dev_1.1-code_review' into dev_1.1

rdaitan committed:
Case-sensitive username authentication

rdaitan committed:
Changed thread body label

rdaitan committed:
Show success page on profile edit

As requested by Mara-san. Show a success page when the user
successfully edited his profile. It has a link to the user's
profile in order to view the changes.

rdaitan committed:
Updated follow updater script

rdaitan committed:
Merge branch 'dev_1.1-code_review' into dev_1.1

rdaitan committed:
Used img_embed.js

rdaitan committed:
Used embed_images.js in comment view as well

rdaitan committed:
Merge branch 'dev_1.1-code_review' into dev_1.1

fatimaPHP commented on app/models/follow.php:
```diff
@@ -0,0 +1,98 @@
+    {
+        $db = DB::conn();
+        $row = $db->row(
+                'SELECT * FROM follow WHERE thread_id=? AND user_id=?',
```
Please fix the indent po

rdaitan replied:
Done po. Thank you. :bow:

rdaitan committed:
Fixed indent

rdaitan committed:
Removed Comment::getOrFail()

Comment::get() would now throw a RecordNotFoundException if it
did not find the comment.

rdaitan committed:
Get last comment in thread with MAX()

Return only the id

Conflicts:
	app/controllers/follow_controller.php
	app/models/comment.php

rdaitan committed:
Merge branch 'dev_1.1-code_review' into dev_1.1
